### PR TITLE
vsphere: add circuit breaker between Jupiter Brain and vSphere

### PR DESCRIPTION
**Pull request by henrikhodne**
I was playing with a Circuit Breaker library and decided to try implementing it here. Might help relieve some pressure on the vSphere API if we have a lot of errors. Not 100% sure about the configuration for when to open the breaker. Right now it's the settings from the example, which is when more than 60% of requests are failing, but that might not work too well when things have been running for a long time, since things need to fail for a long time before the circuit opens.

### CODE REVIEW

File: src/github.com/travis-ci/jupiter-brain/vsphere.go [hunk 1]
```diff
@@ -264,6 +280,16 @@ func (i *vSphereInstanceManager) createClient(ctx context.Context) (*govmomi.Cli
 	}
 
 	client.Client.RoundTripper = vim25.Retry(client.Client.RoundTripper, vim25.TemporaryNetworkError(3))
+	client.Client.RoundTripper = &soapBreakerRoundTripper{
+		rt: client.Client.RoundTripper,
+		cb: gobreaker.NewCircuitBreaker(gobreaker.Settings{
+			Name: "vSphere govmomi",
+			ReadyToTrip: func(counts gobreaker.Counts) bool {
+				failureRatio := float64(counts.TotalFailures) / float64(counts.Requests)
+				return counts.Requests >= 3 && failureRatio >= 0.6
```
Comment (meatballhat) [inline src/github.com/travis-ci/jupiter-brain/vsphere.go]: Let's get these injected from config and test some different settings!

Comment (henrikhodne) [inline src/github.com/travis-ci/jupiter-brain/vsphere.go]: We can do that, but my concern with this is the code and math itself, not so much the numerical constants. I agree that whatever code we end up with, we can probably extract the constants to configs.

Comment (meatballhat) [inline src/github.com/travis-ci/jupiter-brain/vsphere.go]: I think the big TODO item I see coming out of today is having a proper staging deployment of jupiter-brain available for testing.  I'd love to see this get some traffic.

Comment (henrikhodne) [inline src/github.com/travis-ci/jupiter-brain/vsphere.go]: For example, right now, if we have Jupiter Brain chugging along and it's made 1,000 passing requests, and then every request starts erroring, the breaker won't trip until we've had 1,500 requests that error.

Comment (meatballhat) [inline src/github.com/travis-ci/jupiter-brain/vsphere.go]: Yep, I can see us having to be fairly conservative.  The cost of 1500 VMs is much higher than 1500 API requests for build status, for example.

**Comment by henrikhodne** [comment]
We should probably set the flag in the circuit breaker that makes it reset the request count every N requests, which would avoid the whole “wait for thousands of requests before anything happens” thing.

Also, I think we should add a state change callback function that logs whenever the circuit breaker trips or resets.

**Comment by emdantrim** [comment]
this is mega stalesies. how do y'all feel about :merge:in' this?

**Comment by meatballhat** [comment]
:sheep: :it: !

**Pull request merged**